src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1,34 +1,208 @@
+						{
+							_cts.Token.ThrowIfCancellationRequested();
+							string messageType;
+							if (!message.TryGet(Constants.MessageType, out messageType))
+								throw new InvalidOperationException(
+									$"Received replication message without type. All messages must have property {Constants.MessageType}");
+
+							switch (messageType)
+							{
+								case "GetLastEtag":
+									_context.Write(_writer, new DynamicJsonValue
+									{
+										[Constants.Replication.PropertyNames.LastSentEtag] = GetLastReceivedEtag(Guid.Parse(_incomingMessageHeader.SourceDatabaseId), _context),
+										[Constants.MessageType] = "GetLastEtag"
+									});
+									_writer.Flush();
+									break;
+								case "ReplicationBatch":
+
+									BlittableJsonReaderArray replicatedDocs;
+									if (!message.TryGet(Constants.Replication.PropertyNames.ReplicationBatch, out replicatedDocs))
+										throw new InvalidDataException(
+											$"Expected the message to have a field with replicated document array, named {Constants.Replication.PropertyNames.ReplicationBatch}. The property wasn't found");
+
+									using (_context.OpenWriteTransaction())
+									{
+										ReceiveDocuments(_context, replicatedDocs);
+										_context.Transaction.Commit();
+									}
+
+									break;
+								default:
+									throw new InvalidOperationException("Unrecognized replication message type.");
+							}
+						}
+					}
+				}
+				catch (Exception e)
+				{
+					OnFailed(e, this);
+				}
+				finally
+				{
+					_context = null;
+					_stream = null;
+				}
+			}
+	    }
+
+	    public string FromToString => $"from {_connectionInfo.SourceDatabaseName} at {_connectionInfo.SourceUrl} (into database {_database.Name})";
+
+	    private long GetLastReceivedEtag(Guid srcDbId, DocumentsOperationContext context)
+	    {
+		    var dbChangeVector = _database.DocumentsStorage.GetDatabaseChangeVector(context);
+		    var vectorEntry = dbChangeVector.FirstOrDefault(x => x.DbId == srcDbId);
+		    return vectorEntry.Etag;
 	    }
 
-	    public void Dispose()
+		private void ReceiveDocuments(DocumentsOperationContext context, BlittableJsonReaderArray docs)
+		{
+			var dbChangeVector = _database.DocumentsStorage.GetDatabaseChangeVector(context);
+			var changeVectorUpdated = false;
+			var maxReceivedChangeVectorByDatabase = new Dictionary<Guid, long>();
+			foreach (BlittableJsonReaderObject doc in docs)
+			{
+				var changeVector = doc.EnumerateChangeVector();
+				foreach (var currentEntry in changeVector)
+				{
+					Debug.Assert(currentEntry.DbId != Guid.Empty); //should never happen, but..
+
+					//note: documents in a replication batch are ordered in incremental etag order
+					maxReceivedChangeVectorByDatabase[currentEntry.DbId] = currentEntry.Etag;
+				}
+
+				const string DetachObjectDebugTag = "IncomingDocumentReplication -> Detach object from parent array";
+				var detachedDoc = context.ReadObject(doc, DetachObjectDebugTag);
```

Comment:
Aren't we duplicating memory usage here?
Also note that you aren't disposing the objects as you go along, so you are consuming the maximum amount of memory, instead of releasing after each document.